### ISSUE TITLE
training loss: nan

### ISSUE THREAD

**Issue by ipsec** [DHDev0/Stochastic-muzero#2]
Hi Daniel,

I'm trying to run a custom environment (works with muzero) with your Stochastic-muzero version.

After creating a config file (just changing env name in experiment_450_config.json) I'm getting 'training loss: nan' while running.

Have you any idea by?

**Comment by ipsec** [comment]
Somewhere in training your code are getting nan to loss and appending it to  muzero_model.store_loss.

[[tensor(nan), tensor(nan), tensor(0.0637), tensor(nan), tensor(0.0687), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0697), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0662), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0654), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0660), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0651), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0596), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0627), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930), tensor(nan), tensor(0.0642), tensor(nan), tensor(nan), tensor(2.2966), tensor(2.2930)], [tensor(nan), tensor(nan), tensor(0.0636), tensor(nan), tensor(0.0664), tensor(nan), tensor(nan), tensor(2.3036), tensor(2.0611), tensor(nan), tensor(0.0665), tensor(nan), tensor(nan), tensor(2.3016), tensor(2.0647), tensor(nan), tensor(0.0628), tensor(nan), tensor(nan), tensor(2.3011), tensor(2.0642), tensor(nan), tensor(0.0634), tensor(nan), tensor(nan), tensor(2.3006), tensor(2.0679), tensor(nan), tensor(0.0661), tensor(nan), tensor(nan), tensor(2.2996), tensor(2.0715), tensor(nan), tensor(0.0601), tensor(nan), tensor(nan), tensor(2.3031), tensor(2.0671), tensor(nan), tensor(0.0645), tensor(nan), tensor(nan), tensor(2.3021), tensor(2.0728), tensor(nan), tensor(0.0610), tensor(nan), tensor(nan), tensor(2.3026), tensor(2.0682), tensor(nan), tensor(0.0631), tensor(nan), tensor(nan), tensor(2.2986), tensor(2.0748)]]

**Comment by ipsec** [comment]
using loss_type "game" its working.

**Comment by DHDev0** [comment]
Hi ipsec,
By any chance could you share the .json that you use.
Even better could you also share the env so i can reproduce the error or give me some details on the game state(observation) , policy and reward function of the env.

If "using loss_type "game" its working." ,It mean it's coming from self.criterion_value , self.criterion_reward, self.value_afterstate_loss or self.vq_vae_commitment_cost inside general loss.(kldiv) So some of them have negative pred or target in kldiv. ( negative log = nan , 0 log = inf )

From your printed loss, it's coming from: 
k<1
self.criterion_value (nan)
self.criterion_policy (nan)

k=>1
self.criterion_value 
self.criterion_policy (nan)
self.criterion_reward
self.value_afterstate_loss (nan)
self.distribution_afterstate_loss (nan)
self.vq_vae_commitment_cost

I don't understand why you would have a problem with the value or prior ( since number of simulation is 0, you use only the prior) So I would be glad to backtest it on my end to understand where it's coming from.

**Comment by ipsec** [comment]
Hi Daniel,

Thanks by fast answer.
This is my config.


```
{
    "human_demonstration_buffer_builder": {
        "keyboard_map_filename": null,
        "set_default_noop": null,
        "path_to_store_game": null
    },
    "game": {
        "env": "gym_orderbook:TraderFlat-v0",
        "render": null,
        "render_fps": 30
    },
    "random_seed": {
        "np_random_seed": 0,
        "torch_manual_seed": 0,
        "env_seed": 0
    },
    "muzero": {
        "model_structure": "mlp_model",
        "state_space_dimensions": 61,
        "hidden_layer_dimensions": 126,
        "number_of_hidden_layer": 4,
        "k_hypothetical_steps": 10,
        "optimizer": "adam",
        "lr_scheduler": "cosineannealinglr",
        "learning_rate": 0.01,
        "loss_type": "game",
        "num_of_epoch": 1000,
        "device": "cpu",
        "load": false,
        "use_amp": false,
        "scaler_on": false,
        "bin_method": "uniform_bin",
        "bin_decomposition_number": 10,
        "priority_scale": 0.5,
        "rescale_value_loss": 1
    },
    "replaybuffer": {
        "window_size": 500,
        "batch_size": 128,
        "td_steps": 50,
        "game_sampling": "priority",
        "position_sampling": "priority",
        "reanalyse_fraction": 0.0,
        "reanalyse_fraction_mode": "chance"
    },
    "monte_carlo_tree_search": {
        "pb_c_base": 19652,
        "pb_c_init": 1.25,
        "discount": 0.997,
        "root_dirichlet_alpha": 0.25,
        "root_exploration_fraction": 0.25,
        "num_simulations": 0,
        "maxium_action_sample": 2,
        "number_of_player": 1,
        "custom_loop": null
    },
    "gameplay": {
        "limit_of_game_play": 500
    },
    "learning_cycle": {
        "number_of_iteration": 1000,
        "number_of_self_play_before_training": 10,
        "number_of_training_before_self_play": 1,
        "temperature_type": "static_temperature",
        "model_tag_number": 450,
        "verbose": true,
        "number_of_worker_selfplay": 0
    },
    "play_game_from_checkpoint": {
        "model_tag": 450,
        "model_device": "cpu",
        "mcts_with_or_without_dirichlet_noise": true,
        "number_of_monte_carlo_tree_search_simulation": 0,
        "temperature": 0,
        "game_iter": 500,
        "slow_mo_in_second": 0.0,
        "render": null,
        "verbose": true
    }
}

```

I have added you with collaborator access to my env.

**Comment by DHDev0** [comment]
Could you print an example of observation , observation shape, action space, action space shape and give me a bound min and max bound for the reward (also to know if it's discrete or continuous )
Else I will dig into it tomorrow.

**Comment by DHDev0** [comment]
You closed the ticket so I will conclude you were able to fix it. If you need help let me know
